Add HuggingFace Pusher Implementation

deep-diver opened the pull request:
#178 project idea relevant implementation

- [X] Tests pass
- [x] Pre-hook commit pass

casassg commented on tfx_addons/huggingface_pusher/README.md:
````diff
@@ -99,6 +99,11 @@ HFSpaceConfig(
 ## Project Dependencies
 - [tfx](https://pypi.org/project/tfx/)
 - [huggingface-hub](https://pypi.org/project/huggingface-hub/)
+- [Git-LFS](https://git-lfs.github.com/) which could be installed as below
+```shell
+$ curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | sudo bash
+$ sudo apt-get install git-lfs
+```
````
why is this needed?

deep-diver replied:
@casassg 

since the size of model file is large, HuggingFace Hub leverages Git-LFS feature to support it. So, I left a snippet to let people know that they need to include it in their Dockerfile.

deep-diver replied:
so this is not a dependency to use the library itself, but it is a dependency to push things to the remote HuggingFace Hub.

casassg replied:
interesting, is there no way to avoid it from having to be used?

If thats the case, I would recommend on Executor adding a short check that validates wether gitlfs exists and if not fails saying what needs to get added and how (aka point to HF pages)

Guessing HF code will already fail, but may be good to capture the error either before it fails there or capture the exception and parse it correctly for users of the component

deep-diver replied:
sounds reasonable! thanks for the advice, will do that

deep-diver commented:
@casassg 

added a simple function to check git-lfs installed, and added tests for it in the `Executor`

casassg commented on tfx_addons/huggingface_pusher/executor.py:
```diff
@@ -0,0 +1,137 @@
+    if self._is_git_lfs_installed() is not True:
+      raise RuntimeError(
+          "Git-LFS is not installed. "
+          "Git-LFS installation guide can be found in https://git-lfs.github.com/."
```
nitpick: may be better to instead reference HugginFace documentation here instead.

casassg commented on tfx_addons/huggingface_pusher/README.md:
```diff
@@ -0,0 +1,117 @@
+- `repo_url` : model repository URL. It is something like f"https://huggingface.co/{repo_id}/{branch}"
+- `space_url` : space repository URL. It is something like f"https://huggingface.co/{repo_id}"
+
+### HuggingFace Space spacific configurations
```
```suggestion
### HuggingFace Space specific configurations
```

casassg commented on tfx_addons/huggingface_pusher/component_test.py:
```diff
@@ -0,0 +1,40 @@
+    hf_pusher = HFPusher(
+        username="test_username",
+        access_token="test_access_token",
+        repo_name="test_repo_name",
```
nit: add space_config example

casassg commented on tfx_addons/huggingface_pusher/executor_test.py:
```diff
@@ -0,0 +1,71 @@
+    exe = executor.Executor()
+
+    try:
+      exe.Do({}, {"pushed_model": [standard_artifacts.PushedModel()]}, {})
```
```suggestion
      exe.Do({}, {"pushed_model": [standard_artifacts.PushedModel()]}, {})
      assert False
```

casassg commented on tfx_addons/huggingface_pusher/README.md:
````diff
@@ -0,0 +1,117 @@
+```
+HFModelPusher(
+    username: str,
+    hf_access_token: str,
````
is this a private token? If so we may want to insert it in a more protected way...

deep-diver replied:
Can you guide me how to? I was injecting this token via GitHub Secret in Action

deep-diver replied:
or is it enough to add more descriptive information how to pass this token? like GitHub Action.

casassg replied:
There are 2 ways to go about this: 
- Encrypt your token and have a way to decrypt it (see @hanneshapke work -> https://github.com/tensorflow/tfx-addons/tree/main/tfx_addons/message_exit_handler#usage-example-at-digits )
- Using something like GCP SecretManager where you store the Secret and retrieve its value at execution time -> https://cloud.google.com/vertex-ai/docs/pipelines/secret-manager

deep-diver replied:
thank you, so I will write as how `message_exit_handler` achive this goal by optionally letting users to specify `decrypt_fn`.

deep-diver commented:
> is this a private token? If so we may want to insert it in a more protected way...

except this issue, all comments are addressed

deep-diver commented:
@casassg 

addressed your comment. I have re-defined the component spec with `decrypt_fn` to let users optionally specify a function to decrypt `access_token` if the `access_token` is provided in encrypted form.

casassg reviewed (approved):
/lgtm May be to get @sayakpaul to review (btw we need to merge https://github.com/tensorflow/tfx-addons/pull/192 before to get ownership over this folder to you and @sayakpaul )

sayakpaul commented:
Will do it soon (~day).

sayakpaul commented:
/lgtm

sayakpaul commented:
/merge
